// n8n/dist/modules/data-store/data-store.repository.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataStoreRepository = void 0;
const api_types_1 = require("@n8n/api-types");
const di_1 = require("@n8n/di");
const typeorm_1 = require("@n8n/typeorm");
const n8n_workflow_1 = require("n8n-workflow");
const data_store_column_entity_1 = require("./data-store-column.entity");
const data_store_rows_repository_1 = require("./data-store-rows.repository");
const data_store_entity_1 = require("./data-store.entity");
let DataStoreRepository = class DataStoreRepository extends typeorm_1.Repository {
    constructor(dataSource, dataStoreRowsRepository) {
        super(data_store_entity_1.DataStore, dataSource.manager);
        this.dataStoreRowsRepository = dataStoreRowsRepository;
    }
    async createDataStore(projectId, name, columns) {
        if (columns.some((c) => !api_types_1.DATA_STORE_COLUMN_REGEX.test(c.name))) {
            throw new n8n_workflow_1.UnexpectedError('bad column name');
        }
        let dataStoreId;
        await this.manager.transaction(async (em) => {
            const dataStore = em.create(data_store_entity_1.DataStore, { name, columns, projectId });
            await em.insert(data_store_entity_1.DataStore, dataStore);
            dataStoreId = dataStore.id;
            const queryRunner = em.queryRunner;
            if (!queryRunner) {
                throw new n8n_workflow_1.UnexpectedError('QueryRunner is not available');
            }
            const columnEntities = columns.map((col, index) => em.create(data_store_column_entity_1.DataStoreColumn, {
                dataStoreId,
                name: col.name,
                type: col.type,
                index: col.index ?? index,
            }));
            if (columnEntities.length > 0) {
                await em.insert(data_store_column_entity_1.DataStoreColumn, columnEntities);
            }
            await this.dataStoreRowsRepository.createTableWithColumns(dataStoreId, columnEntities, queryRunner);
        });
        if (!dataStoreId) {
            throw new n8n_workflow_1.UnexpectedError('Data store creation failed');
        }
        const createdDataStore = await this.findOneOrFail({
            where: { id: dataStoreId },
            relations: ['project', 'columns'],
        });
        return createdDataStore;
    }
    async deleteDataStore(dataStoreId, entityManager) {
        const executor = entityManager ?? this.manager;
        return await executor.transaction(async (em) => {
            const queryRunner = em.queryRunner;
            if (!queryRunner) {
                throw new n8n_workflow_1.UnexpectedError('QueryRunner is not available');
            }
            await em.delete(data_store_entity_1.DataStore, { id: dataStoreId });
            await this.dataStoreRowsRepository.dropTable(dataStoreId, queryRunner);
            return true;
        });
    }
    async deleteDataStoreByProjectId(projectId) {
        return await this.manager.transaction(async (em) => {
            const existingTables = await em.findBy(data_store_entity_1.DataStore, { projectId });
            let changed = false;
            for (const match of existingTables) {
                const result = await this.deleteDataStore(match.id, em);
                changed = changed || result;
            }
            return changed;
        });
    }
    async deleteDataStoreAll() {
        return await this.manager.transaction(async (em) => {
            const queryRunner = em.queryRunner;
            if (!queryRunner) {
                throw new n8n_workflow_1.UnexpectedError('QueryRunner is not available');
            }
            const existingTables = await em.findBy(data_store_entity_1.DataStore, {});
            let changed = false;
            for (const match of existingTables) {
                const result = await em.delete(data_store_entity_1.DataStore, { id: match.id });
                await this.dataStoreRowsRepository.dropTable(match.id, queryRunner);
                changed = changed || (result.affected ?? 0) > 0;
            }
            return changed;
        });
    }
    async getManyAndCount(options) {
        const query = this.getManyQuery(options);
        const [data, count] = await query.getManyAndCount();
        return { count, data };
    }
    async getMany(options) {
        const query = this.getManyQuery(options);
        return await query.getMany();
    }
    getManyQuery(options) {
        const query = this.createQueryBuilder('dataStore');
        this.applySelections(query);
        this.applyFilters(query, options.filter);
        this.applySorting(query, options.sortBy);
        this.applyPagination(query, options);
        return query;
    }
    applySelections(query) {
        this.applyDefaultSelect(query);
    }
    applyFilters(query, filter) {
        for (const x of ['id', 'projectId']) {
            const content = [filter?.[x]].flat().filter((x) => x !== undefined);
            if (content.length === 0)
                continue;
            query.andWhere(`dataStore.${x} IN (:...${x}s)`, {
                [x + 's']: content.length > 0 ? content : [''],
            });
        }
        if (filter?.name) {
            const nameFilters = typeof filter.name === 'string' ? [filter.name] : filter.name;
            for (const name of nameFilters) {
                query.andWhere('LOWER(dataStore.name) LIKE LOWER(:name)', {
                    name: `%${name}%`,
                });
            }
        }
    }
    applySorting(query, sortBy) {
        if (!sortBy) {
            query.orderBy('dataStore.updatedAt', 'DESC');
            return;
        }
        const [field, order] = this.parseSortingParams(sortBy);
        this.applySortingByField(query, field, order);
    }
    parseSortingParams(sortBy) {
        const [field, order] = sortBy.split(':');
        return [field, order?.toLowerCase() === 'desc' ? 'DESC' : 'ASC'];
    }
    applySortingByField(query, field, direction) {
        if (field === 'name') {
            query
                .addSelect('LOWER(dataStore.name)', 'datastore_name_lower')
                .orderBy('datastore_name_lower', direction);
        }
        else if (['createdAt', 'updatedAt'].includes(field)) {
            query.orderBy(`dataStore.${field}`, direction);
        }
    }
    applyPagination(query, options) {
        query.skip(options.skip ?? 0);
        if (options?.take) {
            query.skip(options.skip ?? 0).take(options.take);
        }
    }
    applyDefaultSelect(query) {
        query
            .leftJoinAndSelect('dataStore.project', 'project')
            .leftJoinAndSelect('dataStore.columns', 'data_store_column')
            .select([
            'dataStore',
            ...this.getDataStoreColumnFields('data_store_column'),
            ...this.getProjectFields('project'),
        ]);
    }
    getDataStoreColumnFields(alias) {
        return [
            `${alias}.id`,
            `${alias}.name`,
            `${alias}.type`,
            `${alias}.createdAt`,
            `${alias}.updatedAt`,
            `${alias}.index`,
        ];
    }
    getProjectFields(alias) {
        return [`${alias}.id`, `${alias}.name`, `${alias}.type`, `${alias}.icon`];
    }
};
exports.DataStoreRepository = DataStoreRepository;
exports.DataStoreRepository = DataStoreRepository = __decorate([
    (0, di_1.Service)(),
    __metadata("design:paramtypes", [typeorm_1.DataSource,
        data_store_rows_repository_1.DataStoreRowsRepository])
], DataStoreRepository);
