// n8n-nodes-base/dist/credentials/Redis.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Redis = void 0;
class Redis {
    name = 'redis';
    displayName = 'Redis';
    documentationUrl = 'redis';
    properties = [
        {
            displayName: 'Password',
            name: 'password',
            type: 'string',
            typeOptions: {
                password: true,
            },
            default: '',
        },
        {
            displayName: 'User',
            name: 'user',
            type: 'string',
            default: '',
            hint: 'Leave blank for password-only auth',
        },
        {
            displayName: 'Host',
            name: 'host',
            type: 'string',
            default: 'localhost',
        },
        {
            displayName: 'Port',
            name: 'port',
            type: 'number',
            default: 6379,
        },
        {
            displayName: 'Database Number',
            name: 'database',
            type: 'number',
            default: 0,
        },
        {
            displayName: 'SSL',
            name: 'ssl',
            type: 'boolean',
            default: false,
        },
    ];
}
exports.Redis = Redis;
