// @n8n/ai-workflow-builder/dist/tools/helpers/progress.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createProgressReporter = createProgressReporter;
exports.reportStart = reportStart;
exports.reportProgress = reportProgress;
exports.reportComplete = reportComplete;
exports.reportError = reportError;
exports.createBatchProgressReporter = createBatchProgressReporter;
function createProgressReporter(config, toolName) {
    const toolCallId = config.toolCall?.id;
    const emit = (message) => {
        config.writer?.(message);
    };
    const start = (input) => {
        emit({
            type: 'tool',
            toolName,
            toolCallId,
            status: 'running',
            updates: [
                {
                    type: 'input',
                    data: input,
                },
            ],
        });
    };
    const progress = (message, data) => {
        emit({
            type: 'tool',
            toolName,
            toolCallId,
            status: 'running',
            updates: [
                {
                    type: 'progress',
                    data: data ?? { message },
                },
            ],
        });
    };
    const complete = (output) => {
        emit({
            type: 'tool',
            toolName,
            toolCallId,
            status: 'completed',
            updates: [
                {
                    type: 'output',
                    data: output,
                },
            ],
        });
    };
    const error = (error) => {
        emit({
            type: 'tool',
            toolName,
            toolCallId,
            status: 'error',
            updates: [
                {
                    type: 'error',
                    data: {
                        message: error.message,
                        code: error.code,
                        details: error.details,
                    },
                },
            ],
        });
    };
    const createBatchReporter = (scope) => {
        let currentIndex = 0;
        let totalItems = 0;
        return {
            init: (total) => {
                totalItems = total;
                currentIndex = 0;
            },
            next: (itemDescription) => {
                currentIndex++;
                progress(`${scope}: Processing item ${currentIndex} of ${totalItems}: ${itemDescription}`);
            },
            complete: () => {
                progress(`${scope}: Completed all ${totalItems} items`);
            },
        };
    };
    return {
        start,
        progress,
        complete,
        error,
        createBatchReporter,
    };
}
function reportStart(reporter, input) {
    reporter.start(input);
}
function reportProgress(reporter, message, data) {
    reporter.progress(message, data);
}
function reportComplete(reporter, output) {
    reporter.complete(output);
}
function reportError(reporter, error) {
    reporter.error(error);
}
function createBatchProgressReporter(reporter, scope) {
    return reporter.createBatchReporter(scope);
}
