// @n8n/permissions/dist/schemas.ee.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.workflowSharingRoleSchema = exports.credentialSharingRoleSchema = exports.projectRoleSchema = exports.teamRoleSchema = exports.personalRoleSchema = exports.assignableGlobalRoleSchema = exports.globalRoleSchema = exports.roleNamespaceSchema = void 0;
const zod_1 = require("zod");
exports.roleNamespaceSchema = zod_1.z.enum(['global', 'project', 'credential', 'workflow']);
exports.globalRoleSchema = zod_1.z.enum(['global:owner', 'global:admin', 'global:member']);
exports.assignableGlobalRoleSchema = exports.globalRoleSchema.exclude([
    'global:owner',
]);
exports.personalRoleSchema = zod_1.z.enum([
    'project:personalOwner',
]);
exports.teamRoleSchema = zod_1.z.enum(['project:admin', 'project:editor', 'project:viewer']);
exports.projectRoleSchema = zod_1.z.enum([...exports.personalRoleSchema.options, ...exports.teamRoleSchema.options]);
exports.credentialSharingRoleSchema = zod_1.z.enum(['credential:owner', 'credential:user']);
exports.workflowSharingRoleSchema = zod_1.z.enum(['workflow:owner', 'workflow:editor']);
