// n8n/dist/controllers/role.controller.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoleController = void 0;
const decorators_1 = require("@n8n/decorators");
const role_service_1 = require("../services/role.service");
let RoleController = class RoleController {
    constructor(roleService) {
        this.roleService = roleService;
    }
    async getAllRoles() {
        const allRoles = await this.roleService.getAllRoles();
        return {
            global: allRoles.filter((r) => r.roleType === 'global'),
            project: allRoles.filter((r) => r.roleType === 'project'),
            credential: allRoles.filter((r) => r.roleType === 'credential'),
            workflow: allRoles.filter((r) => r.roleType === 'workflow'),
        };
    }
    async getRoleBySlug(slug) {
        return await this.roleService.getRole(slug);
    }
    async updateRole(slug, body) {
        return await this.roleService.updateCustomRole(slug, body);
    }
    async deleteRole(slug) {
        return await this.roleService.removeCustomRole(slug);
    }
    async createRole(body) {
        return await this.roleService.createCustomRole(body);
    }
};
exports.RoleController = RoleController;
__decorate([
    (0, decorators_1.Get)('/'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], RoleController.prototype, "getAllRoles", null);
__decorate([
    (0, decorators_1.Get)('/:slug'),
    __param(0, (0, decorators_1.Param)('slug')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", Promise)
], RoleController.prototype, "getRoleBySlug", null);
__decorate([
    (0, decorators_1.Patch)('/:slug'),
    (0, decorators_1.GlobalScope)('role:manage'),
    __param(0, (0, decorators_1.Param)('slug')),
    __param(1, decorators_1.Body),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", Promise)
], RoleController.prototype, "updateRole", null);
__decorate([
    (0, decorators_1.Delete)('/:slug'),
    (0, decorators_1.GlobalScope)('role:manage'),
    __param(0, (0, decorators_1.Param)('slug')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", Promise)
], RoleController.prototype, "deleteRole", null);
__decorate([
    (0, decorators_1.Post)('/'),
    (0, decorators_1.GlobalScope)('role:manage'),
    __param(0, decorators_1.Body),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], RoleController.prototype, "createRole", null);
exports.RoleController = RoleController = __decorate([
    (0, decorators_1.RestController)('/roles'),
    __metadata("design:paramtypes", [role_service_1.RoleService])
], RoleController);
