// n8n-nodes-base/dist/nodes/DataTable/common/selectMany.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSelectFields = getSelectFields;
exports.getSelectFilter = getSelectFilter;
exports.executeSelectMany = executeSelectMany;
const n8n_workflow_1 = require("n8n-workflow");
const constants_1 = require("./constants");
const fields_1 = require("./fields");
const utils_1 = require("./utils");
function getSelectFields(displayOptions, requireCondition = false) {
    return [
        {
            displayName: 'Must Match',
            name: 'matchType',
            type: 'options',
            options: [
                {
                    name: 'Any Condition',
                    value: constants_1.ANY_CONDITION,
                },
                {
                    name: 'All Conditions',
                    value: constants_1.ALL_CONDITIONS,
                },
            ],
            displayOptions,
            default: constants_1.ANY_CONDITION,
        },
        {
            displayName: 'Conditions',
            name: 'filters',
            type: 'fixedCollection',
            typeOptions: {
                multipleValues: true,
                minRequiredFields: requireCondition ? 1 : 0,
            },
            displayOptions,
            default: {},
            placeholder: 'Add Condition',
            options: [
                {
                    displayName: 'Conditions',
                    name: 'conditions',
                    values: [
                        {
                            // eslint-disable-next-line n8n-nodes-base/node-param-display-name-wrong-for-dynamic-options
                            displayName: 'Column',
                            name: 'keyName',
                            type: 'options',
                            // eslint-disable-next-line n8n-nodes-base/node-param-description-wrong-for-dynamic-options
                            description: 'Choose from the list, or specify using an <a href="https://docs.n8n.io/code/expressions/">expression</a>',
                            typeOptions: {
                                loadOptionsDependsOn: [fields_1.DATA_TABLE_ID_FIELD],
                                loadOptionsMethod: 'getDataTableColumns',
                            },
                            default: 'id',
                        },
                        {
                            // eslint-disable-next-line n8n-nodes-base/node-param-display-name-wrong-for-dynamic-options
                            displayName: 'Condition',
                            name: 'condition',
                            // eslint-disable-next-line n8n-nodes-base/node-param-description-missing-from-dynamic-options
                            type: 'options',
                            typeOptions: {
                                loadOptionsDependsOn: ['&keyName'],
                                loadOptionsMethod: 'getConditionsForColumn',
                            },
                            default: 'eq',
                        },
                        {
                            displayName: 'Value',
                            name: 'keyValue',
                            type: 'string',
                            default: '',
                            displayOptions: {
                                hide: {
                                    condition: ['isEmpty', 'isNotEmpty'],
                                },
                            },
                        },
                    ],
                },
            ],
            description: 'Filter to decide which rows get',
        },
    ];
}
function getSelectFilter(ctx, index) {
    const fields = ctx.getNodeParameter('filters.conditions', index, []);
    const matchType = ctx.getNodeParameter('matchType', index, constants_1.ANY_CONDITION);
    const node = ctx.getNode();
    if (!(0, utils_1.isMatchType)(matchType)) {
        throw new n8n_workflow_1.NodeOperationError(node, 'unexpected match type');
    }
    if (!(0, utils_1.isFieldArray)(fields)) {
        throw new n8n_workflow_1.NodeOperationError(node, 'unexpected fields input');
    }
    return (0, utils_1.buildGetManyFilter)(fields, matchType);
}
async function executeSelectMany(ctx, index, dataStoreProxy, rejectEmpty = false) {
    const filter = getSelectFilter(ctx, index);
    if (rejectEmpty && filter.filters.length === 0) {
        throw new n8n_workflow_1.NodeOperationError(ctx.getNode(), 'At least one condition is required');
    }
    let take = 1000;
    const result = [];
    let totalCount = undefined;
    do {
        const response = await dataStoreProxy.getManyRowsAndCount({
            skip: result.length,
            take,
            filter,
        });
        const data = response.data.map((json) => ({ json }));
        // Optimize common path of <1000 results
        if (response.count === response.data.length) {
            return data;
        }
        if (totalCount !== undefined && response.count !== totalCount) {
            throw new n8n_workflow_1.NodeOperationError(ctx.getNode(), 'synchronization error: result count changed during pagination');
        }
        totalCount = response.count;
        result.push.apply(result, data);
        take = Math.min(take, response.count - result.length);
    } while (take > 0);
    return result;
}
